torch/distributed/_functional_collectives.py

```diff
@@ -379,6 +379,37 @@ def all_to_all_single(
     return _maybe_wrap_tensor(tensor)
 
 
+def permute_tensor(
+    self: torch.Tensor,
+    src_dst_pairs: List[Tuple[int, int]],
```

Comment:
Looks great! A few suggestions for your consideration:
- IIUC, each rank can only send to a single peer and receive from a single peer. I believe violating this would lead to undefined behavior (e.g. a rank receives from two peers would cause the buffer to be written simultaneously). Adding checks to verify the constraint would make the API more user friendly.
- To make the API more fool-proof and the input checking logic simpler, maybe consider changing `List[Tuple[int, int]]` -> `List[int]` (e.g. `src_dst[m] == n` means m sends to n).

Reply:
Great suggestion! Was a little nervous about using a list in case only a small amount of ranks in a large world were participating in the op, but I think that's probably an edge case and the list is fine.